File: Programs/Shared/BuildConfig.go
// Copyright 2024 mzoesch. All rights reserved.

package Shared

import (
    "fmt"
    "strings"
)

type BuildConfig int

const (
    BCFG_SHIPPING    BuildConfig = iota
    BCFG_DEVELOPMENT BuildConfig = iota
    BCFG_DEBUG       BuildConfig = iota
)

func (config BuildConfig) ToString() string {
    switch config {
    case BCFG_SHIPPING:
        return "Shipping"
    case BCFG_DEVELOPMENT:
        return "Development"
    case BCFG_DEBUG:
        return "Debug"
    default:
        panic("Unknown build configuration")
    }
}

func (config BuildConfig) IsShipping() bool {
    return config == BCFG_SHIPPING
}

func (config BuildConfig) IsDevelopment() bool {
    return config == BCFG_DEVELOPMENT
}

func (config BuildConfig) IsDebug() bool {
    return config == BCFG_DEBUG
}

func IsValidBuildConfig(config string) bool {
    switch config {
    case "Shipping":
        return true
    case "Development":
        return true
    case "Debug":
        return true
    default:
        return false
    }
}

func GetBuildConfigFromString(config string) BuildConfig {
    if strings.Contains(config, BCFG_DEBUG.ToString()) {
        return BCFG_DEBUG
    }

    if strings.Contains(config, BCFG_DEVELOPMENT.ToString()) {
        return BCFG_DEVELOPMENT
    }

    if strings.Contains(config, BCFG_SHIPPING.ToString()) {
        return BCFG_SHIPPING
    }

    panic(fmt.Sprintf("Unknown build configuration [%s].", config))

    return BCFG_DEBUG
}

File: Programs/Meta/BuildConfig.py
# Copyright mzoesch. All rights reserved.

from enum import Enum


class BuildConfiguration(Enum):
    UNKNOWN = 0
    DEBUG = 1
    DEVELOPMENT = 2
    SHIPPING = 3


def build_configuration_to_string(config: BuildConfiguration) -> str:
    if config == BuildConfiguration.DEBUG:
        return 'Debug'

    if config == BuildConfiguration.DEVELOPMENT:
        return 'Development'

    if config == BuildConfiguration.SHIPPING:
        return 'Shipping'

    return 'Unknown'


def build_configuration_to_define(config: BuildConfiguration) -> str:
    if config == BuildConfiguration.DEBUG:
        return 'IN_DEBUG'

    if config == BuildConfiguration.DEVELOPMENT:
        return 'IN_DEVELOPMENT'

    if config == BuildConfiguration.SHIPPING:
        return 'IN_SHIPPING'

    raise Exception('Unknown build configuration')
